---
title: "Novel Exon classification version 2"
output: html_notebook
---

```{r setup}
knitr::opts_knit$set(root.dir = '~/NIH/')
#knitr::opts_knit$set(root.dir = '/data/swamyvs/eyeintegration_splicing/')
```


```{r}
library(tidyverse)
library(RBedtools)
```


From classify novel exons script, can also load exon classifcation rdata file from pipeline, but its broken atm 
```{r}

gfc_gtf <- rtracklayer::readGFF('occular_transcriptomes_paper/all_tissues.combined.gtf')
load('dev_eyeintegration_splicing/rdata/all_ref_tx_exons.rdata')
novel_transcripts <- anti_join(gfc_gtf %>% filter(type == 'transcript'), all_transcripts) %>% filter(!grepl('TCONS', gene_name))
novel_loci <- anti_join(gfc_gtf %>% filter(type == 'transcript'), all_transcripts) %>% filter(grepl('TCONS', gene_name))
novel_single_exon_tx <- novel_transcripts$transcript_id %>% {filter(gfc_gtf, transcript_id  %in% .)} %>% group_by(transcript_id) %>%
    summarise(count=n()) %>% filter(count == 2) %>% pull(transcript_id) %>%  {filter(novel_transcripts, transcript_id %in% .)}
novel_transcripts <- filter(novel_transcripts, !transcript_id %in% novel_single_exon_tx$transcript_id)



novel_exons <- gfc_gtf %>% 
    filter(type == 'exon',
           !transcript_id %in% novel_loci$transcript_id, !transcript_id %in% novel_single_exon_tx$transcript_id ) %>%
    select(seqid, strand, start, end) %>% 
    anti_join( all_exons) %>%
    mutate(id=paste0('nvl_exon', 1:nrow(.)))

nvl_start <- anti_join(novel_exons %>% select(seqid, strand, start, id), all_exons) %>% { novel_exons$id  %in% .$id}
nvl_end <- anti_join(novel_exons %>% select(seqid, strand, end, id), all_exons) %>% {novel_exons$id  %in% .$id }

novel_exons <- novel_exons %>% mutate(nv_type=case_when(nvl_start & nvl_end ~ 'novel_exon',
                                         !nvl_start & nvl_end ~ 'A3SS',
                                         nvl_start & !nvl_end ~ 'A5SS',
                                         !nvl_start & !nvl_end ~ 'RI'))




all_exon_starts <- gfc_gtf %>% filter(exon_number == 1) %>%  select(seqid,strand, start, end) %>% distinct %>% mutate(is.TSS=T)
all_exon_ends <- gfc_gtf %>% filter(type == 'exon') %>% group_by(transcript_id) %>% 
    summarise(seqid=last(seqid), strand=last(strand), start=last(start), end=last(end)) %>% select(-transcript_id) %>% distinct %>% mutate(is.TES=T)
novel_exons <-  novel_exons %>% left_join(all_exon_starts) %>% left_join(all_exon_ends)
novel_exons[is.na(novel_exons)] <- F
novel_exons <-novel_exons %>% mutate(nv_type_rc= case_when(is.TSS ~ 'TSS', is.TES ~ 'TES', TRUE ~ nv_type))
table(novel_exons$nv_type_rc) %>% print


gfc_gtf_ano <- filter(gfc_gtf, !transcript_id %in% novel_loci$transcript_id)
gfc_gtf_ref <- filter(gfc_gtf_ano, !transcript_id %in% novel_transcripts$transcript_id)

gfc_gtf_full <-  gfc_gtf_ano %>% filter(transcript_id  %in% novel_transcripts$transcript_id) %>% select(seqid, strand, start, end) %>%
    distinct %>% anti_join(gfc_gtf_ref) %>% anti_join(all_exons) %>% anti_join(all_transcripts) %>%  
    mutate(is.novel=T) %>% left_join(gfc_gtf_ano, .) %>% mutate(is.novel=replace_na(is.novel, F))




# 
# uniq_tss <-  gfc_gtf_full %>% filter(exon_number==1) %>% 
#     select(seqid, strand, start) %>% distinct %>% 
```


Unlike the cancer promoter paper, I'm not removing first exons that are also internal exons 

```{R}
same_start <-  gfc_gtf_full %>% filter(exon_number==1) %>%
     select(seqid, strand, start, end, gene_name) %>% distinct %>%  
    group_by(seqid, strand ,start) %>% 
     summarise(count=n(), max_end=max(end), s_gene_name=first(gene_name)) %>% filter( count >1)
# constructing a merged super FE for all distinct Starts
uniq_starts <- gfc_gtf_full %>% filter(exon_number==1) %>%
     select(seqid, strand, start, end, gene_name) %>% distinct %>% 
     anti_join(same_start) %>% 
    bind_rows(., same_start %>% select(seqid, strand, start, end=max_end, gene_name=s_gene_name))

multi_start_genes <- uniq_starts %>% group_by(gene_name) %>% summarise(count=n()) %>% filter(count >1) %>% pull(gene_name)
uniq_start_multi_gene <- novel_exons %>% mutate(novel_start=nvl_start) %>% filter(novel_start) %>% 
    select(seqid, strand, start, novel_start) %>% distinct %>%  
    left_join(uniq_starts, .) %>% mutate(novel_start=replace_na(novel_start, F)) %>% filter(gene_name %in% multi_start_genes)



```


I initially tried to define promoter by binning different regions of the genome and associating promoters to bins ie all exon in a bin > single promoter, but it made sense pretty quickly, bc distance is relative, so went with the overlapping exons method/

```{r}
gdata::keep(uniq_start_multi_gene, novel_exons, gfc_gtf, gfc_gtf_full, all_exons, all_transcripts, sure = T)
trancript_bed <- gfc_gtf %>% mutate(score=999) %>% filter(type == 'transcript') %>% 
    select(seqid, start, end, transcript_id, score, strand) %>% distinct
fe_bed <- uniq_start_multi_gene %>% mutate(score=999, fe_id=paste0('fe_',1:nrow(.))) %>% 
    select(seqid, start, end, fe_id, score, strand)

merged_fe <- fe_bed %>% 
    from_data_frame %>% 
    RBedtools('sort', output = 'stdout',  i=.)  %>%
    RBedtools('merge', options = '-s -c 4,6 -o collapse,distinct', i=.) %>% # stranded merge, report fe ids and strand
    to_data_frame 
colnames(merged_fe) <- c('p_seqid', 'p_start', 'p_end', 'ids', 'p_strand')
# these locations are the merged
make_df <- function(df){
    fe_id <- df %>% pull(ids) %>% str_split(',') %>% .[[1]] %>% {tibble(fe_id=.)}
    n=nrow(fe_id)
    res<- replicate(n,df[1,c('p_seqid', 'p_start', 'p_end', 'p_strand', 'promoter_id')], simplify=F) %>% bind_rows() %>% bind_cols(.,fe_id)
    return(res)
    
}

promoter2fe <- merged_fe %>% mutate(promoter_id=paste0('pro_', 1:nrow(.))) %>% 
     split(1:nrow(.)) %>% 
    lapply(make_df) %>% 
    bind_rows 

#these are all
fe2_tx <- inner_join(promoter2fe, fe_bed %>% select(-score )) %>% 
    inner_join(gfc_gtf_full %>% filter(type == 'exon', exon_number == 1) %>% 
                         select(seqid, strand, start, end, transcript_id, is.novel)) 
                
        
                     

```

```{r}
num_unique_prom <-  fe2_tx %>% group_by(promoter_id) %>% summarise(completly_novel= all(is.novel)) %>%
    pull(completly_novel) %>% sum 
num_nv_TSS <- filter(novel_exons, nv_type_rc=='TSS') %>% pull(id) %>% unique %>% length 
num_unique_prom/num_nv_TSS

novel_prom_ids <- fe2_tx %>% group_by(promoter_id) %>% 
    summarise(completly_novel= all(is.novel)) %>% filter(completly_novel) %>% 
    pull(promoter_id)


novel_promoters <- filter(fe2_tx, promoter_id %in% novel_prom_ids)
ref_promoters <- filter( fe2_tx, !promoter_id %in% novel_prom_ids )
```



Now lets see how promoters overlap with cage peaks




```{r}
library(RBedtools)

cage <- read_tsv('occular_transcriptomes_paper/hg38_CAGE_coords.bed.gz', col_names = F) 
cage_bed <- cage %>% 
    mutate(X4=paste0('cage_', 1:nrow(.))) %>% select(-X7,-X8,-X9) %>% from_data_frame %>% 
    RBedtools('sort', i=.)

novel_promoter_bed <- novel_promoters %>% mutate(score=999,fend=p_start+1) %>% 
    select(p_seqid, p_start, fend, promoter_id, score, p_strand) %>% 
    from_data_frame %>% RBedtools('sort', i=.)

ref_promoter_bed <- ref_promoters %>% mutate(score=999,fend=p_start+1) %>% 
    select(p_seqid, p_start, fend, promoter_id, score, p_strand) %>% 
    from_data_frame %>% RBedtools('sort', i=.)



novel_closest <- RBedtools('closest',options = '-s -D a', a=novel_promoter_bed, b=cage_bed) %>% to_data_frame
ref_closest <- RBedtools('closest',options = '-s -D a', a=ref_promoter_bed, b=cage_bed) %>% to_data_frame

sum(abs(novel_closest$X13) <=100) / nrow(novel_promoters)
sum(abs(ref_closest$X13)<=100) / nrow(ref_promoters)
#maybe try Hyper geometric test?
save.image(file = 'occular_transcriptomes_paper/data/alt_promoter_xploring.Rdata')

```

significant or not?

we can define each promter regions as (tss-100, tss+50)








