---
title: "salmon_bootstrap variance"
output: html_notebook
---

```{r setup}
knitr::opts_knit$set(root.dir = '~/NIH/dev_eyeintegration_splicing/')
#knitr::opts_knit$set(root.dir = '/data/swamyvs/eyeintegration_splicing/')
```


Examine the bootstrap variablity of salmon quantifcation, and use that to inform  transcriptome

These salmon runs comes from the de novo gtf built by stringtie with out any quantification filtering

```{r}
library(tidyverse)
library(matrixStats)
# bs_files <- list.files('data/quant_files/RPE_Fetal.Tissue', pattern = 'quant_bootstraps.tsv.gz', recursive = T, full.names = T)
# 
proc_boostrap <- function(file){
    BS_raw <- read_tsv(file, col_names = F)
    name <- str_split(file,'/')[[1]][4]
    res <- tibble(transcript_id=BS_raw$X1, !!name := BS_raw[,-1] %>% as.matrix %>% {matrixStats::rowVars(.)} )
    return(res)
}
# 
# 
# all_bs <- lapply(bs_files, proc_boostrap) %>% reduce(left_join)
load('rdata/salmon_bs_rpe_fetal_pre_filt.rdata')
var_sum <- function(all_bs){
    novel_bs <- filter(all_bs, grepl('MSTRG', transcript_id))
    ref_bs <- filter(all_bs, !grepl('MSTRG', transcript_id) )
    novel_medvar <- novel_bs[,-1] %>% as.matrix() %>% rowMedians()
    ref_medvar <- ref_bs[,-1] %>% as.matrix %>% rowMedians()
    print('ref')
    summary(ref_medvar) %>% print()
    print('novel')
    summary(novel_medvar) %>% print
    refvar_95 <- quantile(ref_medvar, .95)
    print('% novel tx kept')
    print(sum(novel_medvar <= refvar_95)/ nrow(novel_bs) ) 
    plot <- ggplot() + 
        geom_density(data = tibble(ref=ref_medvar, cond='Reference'), aes(x=ref, color=cond)) +
        geom_density(data=tibble(novel=novel_medvar, cond='Novel'), aes(x=novel, color=cond)) +
        geom_vline(xintercept=refvar_95,colour='green')+
        geom_text(aes(x=refvar_95, y=.005,label='99th Percentile'),angle=90)+
        xlab('Salmon Transcript Variance')+
        xlim(c(0,refvar_95+100)) + 
        ggtitle('Distribution of Quantification Variance for Reference and Novel Transcripts')+
        #ylim(c(0,.005)) + 
        theme_minimal()
    print(plot)
    # plot2 <- ggplot() + 
    #     geom_density(data = tibble(ref=ref_medvar), aes(x=ref), colour='blue') +
    #     geom_density(data=tibble(novel=novel_medvar), aes(x=novel), colour='green') +
    #     xlab('Salmon Transcript Variance')+
    #     xlim(c(0,80)) + 
    #     theme_minimal()   
    # print(plot2)
}

var_sum(all_bs_rpe_tpec)
```

check only transcripts that passed quantification filter

```{r}
filtered_gtf <- rtracklayer::readGFF('data/gtfs/filtered_tissue/RPE_Fetal.Tissue.gtf')# this is the gtf filtered by salmon quantification
full_gtf <- rtracklayer::readGFF('data/gtfs/tissue_gtfs/RPE_Fetal.Tissue_st.gtf')# this is the complete gtf(for RPE_Fetal.Tissue)

all_bs_quan_filt <- filter(all_bs_rpe_tpec, transcript_id %in% filtered_gtf$transcript_id)
all_bs_quan_filt <- filter(all_bs_quan_filt, transcript_id %in% filtered_gtf$transcript_id)
var_sum(all_bs_quan_filt)
```

filtering based on expression makes  novel and ref look more similar

So based on 95%  ref var cutoff,  ~ 25 percent of novel transcripts got too much variance.



I wanted to check how the bootstraps looked from the gtf made by merging all tissues - this merged gtf is made by merging all tissue specifc gtfs, which were filtered only on expression.

```{r}
#rm(list=ls())
load('rdata/all_salmon_bs.rdata')#  ran and saved 
all_gtf <- rtracklayer::readGFF('data/gtfs/all_tissues.combined.gtf')
tcons2ens <- all_gtf %>% filter(type == 'transcript', transcript_id %in% all_bs_qall$transcript_id) %>% select(transcript_id, oId)
all_bs_qall <- all_bs_qall %>% left_join(tcons2ens, .) %>% select(-transcript_id, transcript_id=oId)
var_sum(all_bs_qall)
```

Now the dist's are a lot more similar, and now seems like the ref tx are a lot more variable 

look at only the RPE like before

```{r}
sample_table <- read_tsv('sampleTableDev.tsv', col_names = c('sample','run', 'paired', 'tissue', 'subtissue', 'origin'))
rpe_var <- filter(sample_table, subtissue=='RPE_Fetal.Tissue') %>% pull(sample) %>% {all_bs_qall[,c('transcript_id', .)]}
var_sum(rpe_var)  
```

still looks the same as the all
bottom half of ref is less variable, but still not the same as above 
dists look exactly the same


filter by exp, then re check var


```{r message=FALSE, warning=FALSE}
library(tximport)
# quant_files <- list.files('data/quant_files_all', pattern = 'quant.sf', recursive = T, full.names = T)
# qnames <- str_split(quant_files, '/') %>% sapply(function(x) x[3])
# txi <- tximport(files = quant_files, type = 'salmon', txOut = T, countsFromAbundance = 'lengthScaledTPM')
# colnames(txi$counts) <- qnames
# counts <- txi$counts %>% as.data.frame() %>% mutate(transcript_id=rownames(.)) %>% select(transcript_id, everything())
#save(counts, file = 'rdata/salmon_all_tissues_quant.Rdata')
load('rdata/salmon_all_tissues_quant.Rdata')
sample_table <- read_tsv('sampleTableDev.tsv', col_names = c('sample', 'run', 'paired', 'tissue', 'subtissue', 'origin' ))
rpe_counts <- filter(sample_table, subtissue=='RPE_Fetal.Tissue') %>% pull(sample) %>% {counts[,c('transcript_id', .)]}
counts_filt <- rpe_counts %>% filter(rowSums(.[,-1]) >= (ncol(.)-1) )
rpe_var_filt <- filter(tcons2ens, transcript_id %in% counts_filt$transcript_id) %>% pull(oId) %>% {filter(rpe_var, transcript_id  %in%  . )}
var_sum(rpe_var_filt)
```
So It seems like transcripts that were variable in the tissue specifc quantification are not as variable in the all_tissue quantification



Now compare the transcripts that are getting dropped in 

```{r message=FALSE, warning=FALSE}

which_var_fail <- function(all_bs){
    novel_bs <- filter(all_bs, grepl('MSTRG', transcript_id))
    ref_bs <- filter(all_bs, !grepl('MSTRG', transcript_id) )
    novel_medvar <- novel_bs[,-1] %>% as.matrix() %>% rowMedians()
    ref_medvar <- ref_bs[,-1] %>% as.matrix %>% rowMedians()
    refvar_95 <- quantile(ref_medvar, .95)
    print(refvar_95)
    failed_novel <- filter(novel_bs, novel_medvar > refvar_95)
    return(failed_novel)
}

track_file <- read_tsv('testing/all_tissues.tracking', col_names = F)
tx_ids <- track_file$X1
det <- str_split(track_file$X7, ':|\\|') 
tcons2oid <- lapply(1:length(det), function(i)  det[[i]][-1] %>%  
                                                .[is.na(as.numeric(.))] %>% 
                                                 {tibble(transcript_id=rep(tx_ids[i], length(.)),oId= . )}) %>% bind_rows()

failed_tspec <- which_var_fail(all_bs_quan_filt) %>% rename(oId=transcript_id) %>% inner_join(tcons2oid,.)# these failed in the tissue spec set
failed_allt <- which_var_fail(rpe_var_filt) %>% rename(oId=transcript_id) %>% inner_join(tcons2ens,.) # these failed in all tissue set 
intersect(failed_tspec$transcript_id, failed_allt$transcript_id) %>% length 

```
so only about a quarter of the transcripts are dropped in both sets between the transcripts being dropped. 

```{r}
tissue_spec_bs_filt <-  all_bs_quan_filt %>%  rename(oId=transcript_id) %>% inner_join(tcons2oid,.) %>% 
    mutate(med_var=rowMedians(.[,-(1:2)] %>% as.matrix)) %>% 
    select(transcript_id, oId, med_var)

all_tissue_filt_medvar <- rpe_var_filt %>%  rename(oId=transcript_id) %>% inner_join(tcons2ens,.) %>% 
    mutate(med_var=rowMedians(.[,-(1:2)] %>% as.matrix)) %>% 
    select(transcript_id, oId, med_var)
nrow(tissue_spec_bs_filt)
nrow(all_tissue_filt_medvar)
intersect(tissue_spec_bs_filt$transcript_id, all_tissue_filt_medvar$transcript_id) %>% length()
```
So there's stuff in the 













```{r}
tissue_spec_bs_filt %>% filter(transcript_id %in% failed_allt$transcript_id)
nrow(failed_allt)
```






