---
title: "salmon_bootstrap variance"
output: html_notebook
---

```{r setup}
knitr::opts_knit$set(root.dir = '~/NIH/dev_eyeintegration_splicing/')
#knitr::opts_knit$set(root.dir = '/data/swamyvs/eyeintegration_splicing/')
```


Examine the bootstrap variablity of salmon quantifcation, and use that to inform  transcriptome

These salmon runs comes from the de novo gtf built by stringtie with out any quantification filtering

```{r}
library(tidyverse)
library(matrixStats)
# bs_files <- list.files('data/quant_files/RPE_Fetal.Tissue', pattern = 'quant_bootstraps.tsv.gz', recursive = T, full.names = T)
# 
# proc_boostrap <- function(file){
#     BS_raw <- read_tsv(file, col_names = F)
#     name <- str_split(file,'/')[[1]][4]
#     res <- tibble(transcript_id=BS_raw$X1, !!name := BS_raw[,-1] %>% as.matrix %>% {matrixStats::rowVars(.)} )
#     return(res)
# }
# 
# 
# all_bs <- lapply(bs_files, proc_boostrap) %>% reduce(left_join)
load('rdata/salmon_bs_rpe_fetal_pre_filt.rdata')
var_sum <- function(all_bs){
    novel_bs <- filter(all_bs, grepl('MSTRG', transcript_id))
    ref_bs <- filter(all_bs, !grepl('MSTRG', transcript_id) )
    novel_medvar <- novel_bs[,-1] %>% as.matrix() %>% rowMedians()
    ref_medvar <- ref_bs[,-1] %>% as.matrix %>% rowMedians()
    print('ref')
    summary(ref_medvar) %>% print()
    print('novel')
    summary(novel_medvar) %>% print
    refvar_95 <- quantile(ref_medvar, .95)
    print('% novel tx kept')
    print(sum(novel_medvar <= refvar_95)/ nrow(novel_bs) ) 
    plot <- ggplot() + 
        geom_density(data = tibble(ref=ref_medvar), aes(x=ref), colour='blue') +
        geom_density(data=tibble(novel=novel_medvar), aes(x=novel), colour='green') +
        geom_vline(xintercept=refvar_95,colour='red')+
        xlim(c(0,refvar_95+100)) + 
        #ylim(c(0,.005)) + 
        theme_minimal()
    print(plot)
    plot2 <- ggplot() + 
        geom_density(data = tibble(ref=ref_medvar), aes(x=ref), colour='blue') +
        geom_density(data=tibble(novel=novel_medvar), aes(x=novel), colour='green') +
        xlim(c(0,80)) + 
        theme_minimal()   
    print(plot2)
}

var_sum(all_bs_rpe_tpec)
```

check only transcripts that passed quantification filter

```{r}
filtered_gtf <- rtracklayer::readGFF('data/gtfs/filtered_tissue/RPE_Fetal.Tissue.gtf')# this is the gtf filtered by salmon quantification
full_gtf <- rtracklayer::readGFF('data/gtfs/raw_tissue_gtfs/RPE_Fetal.Tissue_st.gtf')# this is the complete gtf(for RPE_Fetal.Tissue)

all_bs_quan_filt <- filter(all_bs_rpe_tpec, transcript_id %in% filtered_gtf$transcript_id)
all_bs_quan_filt <- filter(all_bs_quan_filt, transcript_id %in% filtered_gtf$transcript_id)
var_sum(all_bs_quan_filt)
```

filtering based on expression makes  novel and ref look more similar

So based on 95%  ref var cutoff,  ~ 25 percent of novel transcripts got too much variance.

lets check how the variance changes once the transcripts are trimmed
(NOTE - I removed the part about looking at the merged tissue quant, becasue theres another notebook that more directly explains the difference in quantification betweent the alltissue merged vs tissue specifc)




```{r}
# 
# proc_boostrap <- function(file){
#     BS_raw <- read_tsv(file, col_names = F)
#     name <- str_split(file,'/')[[1]][4]
#     res <- tibble(transcript_id=BS_raw$X1, !!name := BS_raw[,-1] %>% as.matrix %>% {matrixStats::rowVars(.)} )
#     return(res)
# }
# filt_bs_files <- list.files('data/filter_tx_quant_files/RPE_Fetal.Tissue', patter='quant_bootstraps.tsv', recursive = T, full.names = T)
# rpe_filt_quant_bs <-  lapply(filt_bs_files, proc_boostrap) %>% reduce(left_join)
# save(rpe_filt_quant_bs, file = 'rdata/rpe_filt_quan_bs.rdata')
load('rdata/rpe_filt_quan_bs.rdata')
var_sum(rpe_filt_quant_bs)
```




```{r}

which_var_fail <- function(all_bs){
    novel_bs <- filter(all_bs, grepl('MSTRG', transcript_id))
    ref_bs <- filter(all_bs, !grepl('MSTRG', transcript_id) )
    novel_medvar <- novel_bs[,-1] %>% as.matrix() %>% rowMedians()
    ref_medvar <- ref_bs[,-1] %>% as.matrix %>% rowMedians()
    refvar_95 <- quantile(ref_medvar, .95)
    print(refvar_95)
    failed_novel <- filter(novel_bs, novel_medvar > refvar_95)
    return(failed_novel)
}
failed_raw <- which_var_fail(all_bs_quan_filt)
failed_filt <- which_var_fail(rpe_filt_quant_bs)
intersect(failed_raw$transcript_id, failed_filt$transcript_id) %>% length

failed_raw %>% filter(!transcript_id %in% failed_filt$transcript_id) %>% View


```

So they are about 92% similar now; and the transcripts that dropped of are really close to the border, so lets add it into the agg+ filter scripts







